jwst/coron/smooth_psf.py

```diff
@@ -0,0 +1,78 @@
+"""Replace bad pixels with the median of the surrounding pixel and median fill
+   smooth the psf images.
+ """
+import logging
+import numpy as np
+
+from scipy.ndimage import median_filter
+
+from jwst.datamodels import dqflags
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+def median_fill_value(input_array, input_dq_array, bsize, xc, yc):
+    """
+        Arguments:
+        ----------
+        input_array : ndarray
+            Input array to filter.
+        input_dq_array : ndarray
+            Input data quality array
+        bsize : scalar
+            box size of the data to extract
+        xc: scalar
+           x position of the data extraction
+        xc: scalar
+           y position of the data extraction
+        """
+    # set the half box size
+    hbox = int(bsize/2)
+
+    # Extract the region of interest for the data
+    try:
+        data_array = input_array[xc - hbox:xc + hbox, yc - hbox: yc + hbox]
+        dq_array = input_dq_array[xc - hbox:xc + hbox, yc - hbox: yc + hbox]
+    except IndexError:
+        # If the box is outside the data return 0
+        log.warning('Box for median filter is outside the data.')
+        return 0.
+
+    filtered_array = data_array[dq_array != dqflags.pixel['DO_NOT_USE']]
+    median_value = np.median(filtered_array)
+
+    if np.isnan(median_value):
+        # If the median fails return 0
+        log.warning('Median filter returned N/A setting value to 0.')
+        median_value = 0.
+
+    return median_value
+
+
+def smooth_psf(psf_model, box_size):
+    """ Routine to median smooth the psf images after flagging and replacing the bad pixels.
```

Comment:
If this comment is correct, in terms of the order of operations, I think it's backwards of the approach that the CoronWG folks used in their example notebook. They first performed a simple `median_filter` of the entire image. creating a temporary smoothed image, and then copied individual pixel values from that smoothed image to replace values in the original image that have DQ flags set. From their notebook:

```
   for i in np.arange(nints):
        # Get the data for this integration
        data_i = np.copy(data[i])
        dq_i = np.copy(dq[i])
        
        # Identify the bad pixels of interest based on 'dq_bits' as 1
        dq_i[dq_i & dq_bits == 0] = 0
        dq_i[dq_i != 0] = 1
        
        # Median-smooth the science data
        smoothed = median_filter(data_i, box_size)
        
        # Replace the bad pixels with their corresponding median-smoothed values
        data_i[dq_i != 0] = smoothed[dq_i != 0]
        data_new[i] = data_i
```

Reply:
So only the bad pixels get modified, as opposed to smoothing the entire image.

Reply:
That's correct. The images to be passed along to the rest of the processing (the rest of `align_refs` and subsequent steps) should be the ones that've only had values for bad pixels replaced. Not the ones that've been smoothed over the entire image.

Reply:
And both the psf and target images need to have this bad pixel replacement applied to them, because the process of aligning is done via some cross-correlation like operation (possibly involving FFT's) between each psf image and each target image, so spurious pixel values in either of them will corrupt the alignment results. The badpix-replaced images for both sets should also then get passed along to subsequent steps.

Reply:
The code is updated to replace bad pixels with the median value calculated from a box around the location of the bad pixel. This is applied to both the psf and target image.